File: app/src/main/java/net/ginteam/carmen/manager/PreferencesManager.java
package net.ginteam.carmen.manager;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.android.gms.maps.model.LatLng;
import com.google.gson.Gson;

import net.ginteam.carmen.CarmenApplication;
import net.ginteam.carmen.common.Constants;
import net.ginteam.carmen.model.city.CityModel;

/**
 * Created by Eugene on 12/28/16.
 */

public class PreferencesManager {

    private static PreferencesManager sInstance;

    private SharedPreferences mPreferences;

    private PreferencesManager() {
        mPreferences = CarmenApplication
                .getContext()
                .getSharedPreferences(Constants.PREFERENCES.NAME, Context.MODE_PRIVATE);
    }

    public static PreferencesManager getInstance() {
        if (sInstance == null) {
            sInstance = new PreferencesManager();
        }
        return sInstance;
    }

    public void setUserToken(String token) {
        mPreferences
                .edit()
                .putString(Constants.PREFERENCES.USER_TOKEN, token)
                .apply();
    }

    public String getUserToken() {
        return mPreferences.getString(Constants.PREFERENCES.USER_TOKEN, "");
    }

    public void setCity(String city) {
        mPreferences
                .edit()
                .putString(Constants.PREFERENCES.CITY, city)
                .apply();
    }

    public CityModel getCity() {
        String cityJson = mPreferences.getString(Constants.PREFERENCES.CITY, "");
        return new Gson().fromJson(cityJson, CityModel.class);
    }

    public void setUserLocation(String userLocation) {
        mPreferences
                .edit()
                .putString(Constants.PREFERENCES.USER_LOCATION, userLocation)
                .apply();
    }
//
    public LatLng getUserLocation() {
       String userLocation = mPreferences.getString(Constants.PREFERENCES.USER_LOCATION, "");

        return !userLocation.isEmpty() ? new LatLng(Double.valueOf(userLocation.split(" ")[0]),
                Double.valueOf(userLocation.split(" ")[1])) : null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/manager/PreferencesManager.kt
package net.ginteam.carmen.kotlin.manager

import android.content.Context
import android.util.Log
import com.google.android.gms.maps.model.LatLng
import com.google.gson.Gson
import net.ginteam.carmen.CarmenApplication
import net.ginteam.carmen.kotlin.Constants
import net.ginteam.carmen.kotlin.model.CityModel

/**
 * Created by eugene_shcherbinock on 2/14/17.
 */

interface PreferencesManager {

    var userAccessToken: String
    var userCityModel: CityModel?
    var userLocation: LatLng?
    var isFirstLaunch: Boolean
    var odometer: String

}

object SharedPreferencesManager : PreferencesManager {

    private val mPreferences
            = CarmenApplication
            .getContext()
            .getSharedPreferences(Constants.Preferences.NAME, Context.MODE_PRIVATE)

    override var userAccessToken: String
        get() = mPreferences.getString(Constants.Preferences.USER_TOKEN, "")
        set(value) {
            mPreferences.edit().putString(Constants.Preferences.USER_TOKEN, value).apply()
        }

    override var userCityModel: CityModel?
        get() {
            val cityJson = mPreferences.getString(Constants.Preferences.CITY, "")
            Log.d("SharedPrefsManager", "Get json model: $cityJson")
            return Gson().fromJson(cityJson, CityModel::class.java)
        }
        set(value) {
            val cityJson = Gson().toJson(value)
            Log.d("SharedPrefsManager", "Set json model: $cityJson")
            mPreferences.edit().putString(Constants.Preferences.CITY, cityJson).apply()
        }

    override var userLocation: LatLng?
        get() {
            val locationString = mPreferences.getString(Constants.Preferences.USER_LOCATION, "")
            val location: LatLng? = if (locationString.isEmpty()) {
                null
            } else {
                LatLng(
                        locationString.split(" ")[0].toDouble(),
                        locationString.split(" ")[1].toDouble()
                )
            }
            return location
        }
        set(value) {
            var locationString = ""
            if (value != null) {
                locationString = String.format("%s %s", value.latitude, value.longitude)
            }
            mPreferences.edit().putString(Constants.Preferences.USER_LOCATION, locationString).apply()
        }

    override var isFirstLaunch: Boolean
        get() = mPreferences.getBoolean(Constants.Preferences.LAUNCH, true)
        set(value) {
            mPreferences.edit().putBoolean(Constants.Preferences.LAUNCH, value).apply()
        }

    override var odometer: String
        get() = mPreferences.getString(Constants.Preferences.ODOMETER, "")
        set(value) {
            mPreferences.edit().putString(Constants.Preferences.ODOMETER, value).apply()
        }
}